std/variant.d

```diff
@@ -2145,6 +2145,61 @@ unittest
     assert(which == -1);
 }
 
+/**
+ * Behaves as $(D visit) but use a function object instead of visiting functions.
+ *
+ * The visiting function object should cover all of variant element types,
+ * and it's statically checked.
+ *
+ * Returns: The return type of visitWithObj is deduced from the visiting function object and
+ * must be the same across all overloads.
+ */
+public auto visitWithObj(VariantType, Handler)(VariantType variant, Handler handler)
+    if (isAlgebraic!VariantType)
+{
+    alias AllowedTypes = VariantType.AllowedTypes;
+
+    // statically check the handler can handler all of variant element types or not.
+    foreach (T; AllowedTypes) {
+        static assert(__traits(compiles, handler(T.init)),
+                Handler.stringof ~ " does not cover " ~ T.stringof);
+    }
+
+    foreach (T; AllowedTypes) {
+        if (T* ptr = variant.peek!T)
```

Comment:
also, I'm confused by the statements of @DmitryOlshansky and @JackStouffer as well. Isn't `variant.type` only known at runtime? `peek` takes a compile-time type. I think you need a static `foreach` over the types -- this is exactly what `visit` does IIRC.